### DISCUSSION TITLE
start local chain with docker-compose

### DISCUSSION THREAD

**Discussion by dckc** [Agoric/agoric-sdk#8480] [Ideas]
@toliaqat recently suggested this, and the idea has been growing on me.

As much as I'd like to us meet the `npm install` market expectations for installing the SDK ([#3857](https://github.com/Agoric/agoric-sdk/issues/3857)), I think much of our near-term audience is likely to have some exposure to node operations and docker. Installing `agd` from npm ([#8033](https://github.com/Agoric/agoric-sdk/issues/8033)) is a good goal to help JS devs, but it looks like sort of a lot of work, and I think it's useful to provide a tool for those familiar with `docker-compose`.

**Comment by dckc** [comment]
## Related Work


We have some experience supporting `docker-compose`:
 - [Setting up an Agoric Dapp Client with docker compose · Agoric/agoric\-sdk Wiki](https://github.com/Agoric/agoric-sdk/wiki/Setting-up-an-Agoric-Dapp-Client-with-docker-compose)

But that's for `ag-solo` as a client of devnet.

We also have [instagoric](https://github.com/Agoric/instagoric) but I don't get the impression that's the sort of thing you can start in 2 or 3 minutes.

We support `agoric start` as a way to start a local chain, but in practice we don't use it much because it doesn't fund the provision pool, set up vaults governance, etc. For some time we tended to use [scenario2 in the cosmic-swingset Makefile](https://github.com/Agoric/agoric-sdk/blob/master/packages/cosmic-swingset/Makefile) but much of the related effort moved to the [[Dockerized Chain Upgrade Tester](https://github.com/Agoric/agoric-sdk/tree/master/packages/deployment/upgrade-test#dockerized-chain-upgrade-tester)](https://github.com/Agoric/agoric-sdk/tree/master/packages/deployment/upgrade-test#dockerized-chain-upgrade-tester), which is now migrating to its own [agoric-3-proposals](https://github.com/Agoric/agoric-3-proposals) repo.

**Comment by mhofman** [comment]
For a local chain we can rely on prebuilt images, but my concern is about requiring docker for dapp developers. The original comment talks about "node operators". How does a "local chain" intersect with "node operators" ?

**Reply by dckc** [reply]
fair point. Maybe what I meant was:

In the short term, much of our audience is exposed to node operations as well as JD dev norms.
For example, [NEAR docs for Creating a Local Development Environment](https://docs.near.org/develop/testing/kurtosis-localnet) say:
> Kurtosis has created an easy way to spin up a local NEAR testing environment using a Docker container.

**Comment by dckc** [comment]
## Prototype boots

Presuming we publish the results of the prerequisite `make build` step, this kinda worked on the first try:

 - [docker-compose.yml](https://github.com/Agoric/agoric-3-proposals/blob/dc-local-chain-docker-compose/docker-compose.yml)

It's small enough to just copy it here:

```yml
version: '3.5'

services:
  agd:
    image: agoric/upgrade-test:agoric-upgrade-11
    ports:
      - 26656:26656
      - 26657:26657
      - 1317:1317
    environment:
      DEST: 1
      DEBUG: "SwingSet:ls,SwingSet:vat"
    volumes:
      - .:/workspace
    entrypoint: /usr/src/agoric-sdk/upgrade-test-scripts/start_to_to.sh
```

To run it: `docker-compose up -d`.

To take a look at logs: `docker-compose logs|less -R`.

### `agd` client commands

You can make `agd` an alias for running `agd` from the container:

```sh
alias agd="docker-compose exec agd agd"
```

Then `agd status` and the like will work, as well as...

```console
$ agd query vstorage children published.priceFeed
children:
- ATOM-USD_price_feed
```

## Goal: simple game dapp

Next step: try to run the [simple game dapp](https://github.com/agoric-labs/dapp-join-game)

cc @sufyaankhan @kbennett2000 @0xpatrickdev @turadg @samsiegart

**Comment by dckc** [comment]
@0xpatrickdev and I got the simple game dapp working yesterday. Here's an idea for office hours:

 1. start a local chain with docker-compose (#8480)
    - [ ] presumes publication of a docker image that you currently have to build locally
 3. Note [gameAssetContract.js](https://github.com/Agoric/agoric-sdk/blob/agoric-upgrade-11wf/packages/smart-wallet/test/gameAssetContract.js) is ~50LOC. `agoric run ...` to build bundles. Install contract bundles using [Gov Proposal Hub](https://agoric-gov-proposal-hub.netlify.app/)
    - [ ] small diversion: open a vault to mint enough IST to install the bundle
    - [ ] 2nd bundle is a bit awkward; here's hoping for other ways to link modules in core-eval proposals
 4. Propose core-eval with Gov Proposal Hub
    - [ ] small diversion: add deposit with CLI, since it's not yet integrated
 5. simulate community voting
    - [ ] currently: using voteLatestProposalAndWait CLI testing utility, until perhaps the Gov Proposal Hub adds the feature
 6. start dapp ui with `yarn; yarn dev`. Note simple dapp UI is easy to build:
    - scaffold: `yarn create vite simple-game --template react-ts`
    - add query support with `@agoric/rpc` and chain watchers - ~100LOC
    - add offer support similarly - ~100LOC


_p.s. #8489 has office hours notes, recording._

**Comment by dckc** [comment]
I got this more integrated with the simple game dapp.

For a recording, see:
https://github.com/agoric-labs/dapp-join-game/pull/2#issuecomment-1791712425

**Comment by dckc** [comment]
Several of us got a bunch more experience with this approach this week. My latest notes are in https://github.com/agoric-labs/gimix/blob/main/README-local-chain.md

I should probably integrate that into dapp-game-pieces.

**Comment by dckc** [comment]
[Getting Started](https://docs.agoric.com/guides/getting-started/) now uses this approach, though the details have changed a bit.
Note the relevant [docker-compose.yml](https://github.com/Agoric/dapp-offer-up/blob/main/docker-compose.yml) is in [Agoric/dapp\-offer\-up](https://github.com/Agoric/dapp-offer-up/tree/main).
